extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java

```diff
@@ -0,0 +1,91 @@
+package io.quarkus.oidc.redis.token.state.manager.runtime;
+
+import static io.quarkus.oidc.runtime.CodeAuthenticationMechanism.SESSION_MAX_AGE_PARAM;
+
+import java.time.Instant;
+import java.util.function.Function;
+import java.util.random.RandomGenerator;
+
+import io.quarkus.oidc.AuthorizationCodeTokens;
+import io.quarkus.oidc.OidcRequestContext;
+import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.oidc.TokenStateManager;
+import io.quarkus.redis.datasource.ReactiveRedisDataSource;
+import io.quarkus.redis.datasource.value.SetArgs;
+import io.quarkus.security.AuthenticationCompletionException;
+import io.quarkus.security.AuthenticationFailedException;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+final class OidcRedisTokenStateManager implements TokenStateManager {
+
+    private final ReactiveRedisDataSource dataSource;
+    private final RandomGenerator randomGenerator;
+
+    OidcRedisTokenStateManager(ReactiveRedisDataSource dataSource) {
+        this.dataSource = dataSource;
+        this.randomGenerator = RandomGenerator.SplittableGenerator.of("L32X64MixRandom");
+    }
+
+    @Override
+    public Uni<String> createTokenState(RoutingContext event, OidcTenantConfig oidcConfig, AuthorizationCodeTokens tokens,
+            OidcRequestContext<String> requestContext) {
+        return createTokenState(AuthorizationCodeTokensRecord.of(tokens), 0, newSetArgs(event))
+                .onFailure().transform(AuthenticationFailedException::new);
+    }
+
+    @Override
+    public Uni<AuthorizationCodeTokens> getTokens(RoutingContext routingContext, OidcTenantConfig oidcConfig, String tokenState,
+            OidcRequestContext<AuthorizationCodeTokens> requestContext) {
+        return dataSource.value(AuthorizationCodeTokensRecord.class).get(tokenState)
+                .onItem().ifNotNull().transform(AuthorizationCodeTokensRecord::toTokens)
+                .onFailure().transform(AuthenticationCompletionException::new);
+    }
+
+    @Override
+    public Uni<Void> deleteTokens(RoutingContext routingContext, OidcTenantConfig oidcConfig, String tokenState,
+            OidcRequestContext<Void> requestContext) {
+        return dataSource.key(String.class).del(tokenState).onFailure().recoverWithNull().replaceWithVoid();
+    }
+
+    private Uni<String> createTokenState(AuthorizationCodeTokensRecord tokenRecord, int attemptCount, SetArgs setArgs) {
+        if (attemptCount >= 3) {
+            return Uni.createFrom().failure(
+                    new RuntimeException("Failed to store OIDC token state in Redis as generated key already existed"));
+        }
+        final String key = "oidc-redis-token-state-" + Instant.now().getEpochSecond() + "-" + randomGenerator.nextInt();
```

Comment:
I know it's weird, but... Redis use `:` as key separator.
So, I believe it should be something like `oidc:redis-token-state`

Also, any reason to use a time and a random id?

Reply:
AFAIK the `:` is only of historical significance, although `oidc:token` seems like a much better prefix than `oidc-redis-token-state`.

When it comes to using a timestamp and a random number, wouldn't a UUID be a much better choice?

Reply:
> So, I believe it should be something like oidc:redis-token-state
> AFAIK the : is only of historical significance, although oidc:token seems like a much better prefix than oidc-redis-token-state.

Alright, thank you, I didn't know about this. I think `oidc:token` is just fine. I'll update it.

> Also, any reason to use a time and a random id?

For random part see answer below. The time is just a cheap way to ensure that generated id doesn't already exist. I'd like to keep it.

> When it comes to using a timestamp and a random number, wouldn't a UUID be a much better choice?

`java.util.UUID#randomUUID` uses secure random. And the secure random was slower based on what I read. TBH I read it on Baeldung https://www.baeldung.com/java-17-random-number-generators#performance (see that table) so I don't know if that is true. I thought it's better to retry in that unlikely situation when random number already existed and use quicker algorithm.

Now, you could say that it attackers could guess random number if they know time range when this was requested and they got issues token state id before and after.

I suppose you would be right, I'll change it to `randomUUID`. thanks @Ladicek

Reply:
I didn't even think of a possible attack vector, I don't understand security that well :-) I was just thinking that UUID exists exactly for this purpose, so why not use it. Performance would be a concern if this was called thousands time per second, but that shouldn't be the case here I think?

Reply:
> Performance would be a concern if this was called thousands time per second, but that shouldn't be the case here I think?

It shouldn't be frequent operation. I always think about performance about I didn't realize compelling reasons for using secure random. I've update it, please check if you want.